### PR TITLE
Implement quantization support

### PR DESCRIPTION
**Pull request by monatis**
It's working with q4_1, but I'd like to also experiment with other quantization types and benchmark performance before mergin it to the master.

### CODE REVIEW

File: models/quantize.cpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+    }
+
+    fin.close();
+    fout.close();
+
+    return true;
+}
+
+// usage:
+//  ./bin/quantize models/ggml-model-f16.bin models/ggml-model-14_1.bin type
```
Comment (Green-Sky) [inline models/quantize.cpp:332]: 14_1 -> q4_1

Comment (monatis) [reply models/quantize.cpp]: Oh thanks :D

**Comment by monatis** [comment]
If I'm not mistaken too badly, quantization has a strange effect in zero shot. q4_0 seems to score negative labels lower --which should be better. I initially thought it would score negatives higher. And q4_0 is only 85.6 MB 😀 

| Q-type | a red apple | a dog | a banana |
- --------- | --------- | --------- | --------- |
| f16 | 0.345 | 0.125 | 0.192 |
| q4_1 | 0.325 | 0.127 | 0.190 |
| q4_0 | 0.339 | 0.109 | 0.187 |

**Comment by Green-Sky** [comment]
I guess we need a better way of testing the quality of the model

**Comment by monatis** [comment]
Agree. Trying to decide on the test data.

**Comment by monatis** [comment]
But anyway, I guess we can merge this and benchmarking might be another PR. In the current state of affairs, we varified that f16, q4_0 and q4_1 can be loaded and do inference in a comparable performance. The next PR can focus on calculating a useful metric on a diverse set of data.

**Comment by Green-Sky** [comment]
took a look using zsl
| text | f16 | q4_0 |
|---|-----|------|
| apple | 0.1595 | 0.1603 | 
| red | 0.1458 | 0.1434 |
| photo | 0.1437 | 0.1440 |
| drawing | 0.1415 | 0.1410 |
| human | 0.1395 | 0.1396 |
| blue | 0.1390 | 0.1394 |
| dog | 0.1311 | 0.1324|

note that the order of `red` and `photo` was switched for the q4_0 quantized model

also, error seems to go into both directions. :)

**Comment by monatis** [comment]
Yes it seems to be more versatile than a single photo and a bunch of labels. I'll design a benchmark hopefully soon. Needed for both our development and announcing the performance properly.

**Pull request merged**